S390x binaries

AlekseiNikiforovIBM opened the pull request:
Allow building nightly, rc and release binaries for s390x.

This PR implements building binaries, but publishing part is currently missing.

AlekseiNikiforovIBM committed:
Build binaries for s390x

malfet commented on .github/s390x-ci/self-hosted-builder/fs/usr/bin/actions-runner:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Nit (I've seen system where bash is in usr/bin rather than bin, but I've yet to see a Linux where env is not in usr/bin)
```suggestion
#!/usr/bin/env bash
```

AlekseiNikiforovIBM replied:
Thanks, I'll update this

malfet commented on .github/s390x-ci/self-hosted-builder/fs/usr/bin/entrypoint:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Is there a reason not to do it in Python? (My bash foo is bad, @seemethere may be you can give there a hand)

AlekseiNikiforovIBM replied:
I took this part from my colleague as is:
https://github.com/zlib-ng/zlib-ng/blob/develop/arch/s390/self-hosted-builder/fs/usr/bin/entrypoint

No other reason to keep it in bash.

malfet commented on .github/workflows/s390x-build-binaries.yml:
```diff
@@ -0,0 +1,50 @@
+jobs:
+  build:
+    runs-on: [linux.s390x]
+    timeout-minutes: 720
```
12 hours?!?! Does it really take that long? Can it reasonably limited to 4-5 hours?

AlekseiNikiforovIBM replied:
Yeah, it was taken from my CI workflow unchanged. Much lower value should be enough here since no tests are running.

malfet reviewed (changes requested):
Looks like this PR is trying to accomplish two things:
 - Enable local cross-build
 - Enable native build using s390 runner

Can we split it in two? I have no objections with either, but it feels wrong to put scripts for local build in `.github/self-hosted-builder` folder, but it's fine to have them in scripts.

Also, how you've tested it so far? (Adding ciflow/nightly tag to see if the workflow is actually invoked)

AlekseiNikiforovIBM commented:
> Looks like this PR is trying to accomplish two things:
> 
>     * Enable local cross-build
> 
>     * Enable native build using s390 runner
> 
> 
> Can we split it in two? I have no objections with either, but it feels wrong to put scripts for local build in `.github/self-hosted-builder` folder, but it's fine to have them in scripts.
> 
> Also, how you've tested it so far? (Adding ciflow/nightly tag to see if the workflow is actually invoked)

This PR is only enabling building binaries for s390x. The builder scripts are only commited to allow anyone else to reproduce build environment on the workers.

Sure, I can split it in two PRs.

I did test on my copy of repository that workflows actually run with copy of following PR:
https://github.com/pytorch/pytorch/pull/113556

But I did not test that this workflow actually triggers although it is different from test workflows.

AlekseiNikiforovIBM committed:
Satisfy lintrunner-noclang

huydhn commented:
@pytorchbot rebase

AlekseiNikiforovIBM committed:
Skip some preparation steps on s390x

These steps are unnecessary on linux.s390x runners,
and sometimes even impossible due to missing configuration,
like aws access.

AlekseiNikiforovIBM committed:
Use publicly available s390x/alpine docker image for s390x

AlekseiNikiforovIBM committed:
Use standard pip repository for s390x

AlekseiNikiforovIBM commented:
@pytorchbot rebase

AlekseiNikiforovIBM commented:
@huydhn  @malfet could you please take a look at this PR? I think it's ready for review and merging.

huydhn commented:
@pytorchbot rebase

AlekseiNikiforovIBM committed:
Regenerate workflow

huydhn commented on .github/actions/test-pytorch-binary/action.yml:
```diff
@@ -2,6 +2,12 @@ name: Test pytorch binary
 
 description: Pulls the docker image and tests the pytorch binary using it. All env variable referenced in the "Test PyTorch binary" step must be set in the GITHUB_ENV file
 
+inputs:
```
I don't think you need this input because I can see `BUILD_ENVIRONMENT` env variable is there.  If you need to use it on the workflow, the syntax is `env.BUILD_ENVIRONMENT`, for example `if: always() && env.BUILD_ENVIRONMENT == 'linux-s390x-binary-manywheel'`

huydhn reviewed (approved):
LGTM! I have one minor comment about the possible redundant of the new `build_environment` inputs to test-pytorch-binary action

AlekseiNikiforovIBM commented:
> LGTM! I have one minor comment about the possible redundant of the new `build_environment` inputs to test-pytorch-binary action

Thanks, I'll get rid of it and rerun CI.

huydhn commented:
@pytorchbot merge
